### ISSUE TITLE
Double scrollbar in Firefox

### ISSUE THREAD

**Issue by meier-andersen** [pipwerks/PDFObject#263]
When using forced PDF.js with Firefox, a double scrollbar appears on the right. It seems that the iframe overflows, and it can be removed by setting the height to 99% 

You can see it in your own example: https://pdfobject.com/examples/pdfjs-forced.html
But only for Firefox 
Another solution could be to set the pdf div (the one with pdfobject-container) to have overflow: hidden

**Comment by pipwerks** [comment]
Hello

I use Firefox on Mac as my primary browser. I don't see the issue you describe.
<img width="665" alt="PDFObject Firefox" src="https://user-images.githubusercontent.com/300566/155769941-74a2526a-17c2-4d2a-9745-f604bef7bd40.png">

What operating system and version of Firefox are you using?

**Comment by meier-andersen** [comment]
@pipwerks I apologize. I should that mentioned that I work on a Windows computer. 
We have replicated it on multiple computers

**Comment by dgw** [comment]
It is not limited to Firefox. I see the same behavior on Chrome 100.0.4896.75 (Win64).

Looking closely it appears there is some whitespace at the end of the iframe, after the viewer itself (see second screenshot).

![image](https://user-images.githubusercontent.com/164140/162051171-91b59992-1d1c-4b3e-9aa1-179cf4b15dd8.png)

![image](https://user-images.githubusercontent.com/164140/162051341-cd6ad70e-112c-4337-bdb8-522a26b9af7b.png)

**Comment by TheArKaID** [comment]
I could confirm that in Firefox 99.0 (Win11 64-bit).

![image](https://user-images.githubusercontent.com/35050722/164322809-eb6085db-39bd-4058-aaaa-ffcda3c77cd6.png)

**Comment by theodore-s-beers** [comment]
So this is only with forced PDF.js, only in Firefox, and only on Windows?

**Comment by dgw** [comment]
> this is only with forced PDF.js

Maybe.

> only on Windows

So far, I guess.

> only in Firefox

Not. Reproduced by me and @TheArKaID on Chrome too.

**Comment by theodore-s-beers** [comment]
> > only in Firefox
> 
> Not. Reproduced by me and @TheArKaID on Chrome too.

Ah, sorry. Should have had another coffee before scanning the thread.

**Comment by theodore-s-beers** [comment]
I found what may be the source of this problem. In the `generatePDFJSMarkup` function, instead of placing the iframe directly within the target node, an intermediate div is added. I don't know why this is done. (Maybe there was a legacy browser compatibility concern?) In the analogous `generatePDFObjectMarkup` function (i.e., when not using PDF.js), the embed/iframe is simply appended as a child of the target node. In my limited testing, removing the extra div in `generatePDFJSMarkup` doesn't cause any problems, and in fact it solves this one. You can see a basic comparison here:

https://www.theobeers.com/pdfjs-test/old/
https://www.theobeers.com/pdfjs-test/new/

Anyway, as things currently stand, when you force PDF.js, you end up with an extra layer in the document tree. You have the `pdfobject-container` div; an intermediate div to which no class name or id is assigned; and the `pdfobject` iframe itself.

The intermediate div is, by default, styled as follows:

```css
position: absolute; top: 0; right: 0; bottom: 0; left: 0;
```

Also interestingly, the target node (i.e., the div that gets the class `pdfobject-container`) has the following styles added to it (*only* when using PDF.js):

```js
targetNode.style.position = "relative";
targetNode.style.overflow = "auto";
```

Again, the necessity of all this isn't clear to me. But there seems to be a problem, in Windows at least, owing to the new div and/or the extra styling. If you load the example page linked by OP; open dev tools; disable the aforementioned styles (on both the target node and the intermediate div); and then add `height: 100%` to the intermediate div; you should see that the double scrollbar is corrected.

This issue could be patched in PDFObject. I guess I would recommend making `generatePDFJSMarkup` operate more like `generatePDFObjectMarkup`. (But I do wonder if there was some reason for the difference…) I'm not sure if there's a workaround in the mean time. You might think that passing the option `omitInlineStyles: true` when calling PDFObject would be helpful—but then you're still stuck with the extra div, and how are you supposed to style it without a class or id? Maybe you could figure something out. I hope this gives a bit of insight into the problem.

**Comment by dgw** [comment]
That's some neat detective work, @theodore-s-beers.

Next question is why this appeared to only affect sites starting late last year or early this one. The anonymous enclosing div you mentioned has been around [since PDFObject 2.0, 6 years ago](https://github.com/pipwerks/PDFObject/commit/0634428640543d6bcb0b99fdad4fc8068e3f4026#diff-b4e47b3ee43837e404abb11a6591aa4655e0006b81f1d1994a6864fd878d28cfR141). [The same goes](https://github.com/pipwerks/PDFObject/commit/0634428640543d6bcb0b99fdad4fc8068e3f4026#diff-b4e47b3ee43837e404abb11a6591aa4655e0006b81f1d1994a6864fd878d28cfR142-R144) for the styles appended to the `targetNode` (same commit). And yet, when implementing a site using PDFObject and forced PDF.js early last year, I didn't see this behavior until revisiting that site to add a new PDF this month. I even have timestamped screenshot proof from internal work chat that the PDF.js embed didn't always have this extra scrollbar.

**New development: Reproduced on Linux** 🎊🎆🎊

Taking advantage of the fact that I had a Linux VM sitting around from… I don't remember when/why (😅)… I tested the "PDF.js (forced)" example page under Ubuntu 20.04 using:

* Chrome 93 (not affected)
* Firefox 92 (affected; see below)
* Chrome 100 (affected; see below)
* Firefox 99 (affected; see below)

These were the versions already installed, and then the versions I got after letting Software Updater updated what it wanted. Seems we at least have a rough idea of where in Chrome's release history the change (can we call this a regression?) happened.

**Additional Repro Screenshots:**

<details>
<summary>Firefox 92 on Ubuntu 20.04</summary>

![image](https://user-images.githubusercontent.com/164140/164568761-0dc90042-97e0-4f12-aa7b-30b381ab5c73.png)

</details>

<details>
<summary>Chrome 100 on Ubuntu 20.04</summary>

![image](https://user-images.githubusercontent.com/164140/164569202-8cc0f56a-e4b9-48a6-8d07-3b4b38640f20.png)

</details>

<details>
<summary>Firefox 99 on Ubuntu 20.04</summary>

![image](https://user-images.githubusercontent.com/164140/164569390-6710289a-b92c-42b3-9c68-9a3cda1b4ae4.png)

</details>

**Comment by dgw** [comment]
Back again with more non-Windows repros: Android 11 (Chrome & Firefox) and iPadOS (Safari, unknown version; not my device).

Mobile devices generally don't display the scrollbars unless the user is actively manipulating that frame, but it's still possible to see weirdness at the top and bottom of the document, or by dragging on the PDF.js UI directly.

It's harder to tell that this bug exists on iPadOS. The way scrolling works hides it pretty well unless you explicitly drag on something that's part of the PDF.js UI and not the document. A Mac user will have to comment whether the same scrolling behavior exists on the desktop OS.

<details>
<summary>Android 11 Chrome</summary>

https://user-images.githubusercontent.com/164140/164575429-4886ab5c-6a7e-4e15-8e4b-fca71f5a5772.mp4

</details>

<details>
<summary>Android 11 Firefox</summary>

https://user-images.githubusercontent.com/164140/164575500-29ddd651-695a-4ec3-b1fb-70d6f73680ac.mp4

</details>

<details>
<summary>iPadOS Safari</summary>

https://user-images.githubusercontent.com/164140/164576092-876be0b6-d64e-4197-b5b4-72aea24e91a0.mp4

</details>

**Comment by theodore-s-beers** [comment]
@dgw good point. I tested with a couple of older versions of Firefox on Windows—as far back as 77, which dates to June 2020—and I still see the issue.

Much more importantly, I have realized thanks to your comment about iPadOS that we *do* have the double scrollbar on macOS, in Chrome, Firefox, and Safari. That's right…  every major browser engine. If you move around in the scrollbar area, you can tell that there are in fact two that can be triggered. It's just more difficult to see because of how scrollbars usually work in macOS when a mouse isn't connected. (Geez, are that many of us working on MacBooks with trackpads these days? It's the one hardware-software combo that masks this problem. Sort of an unsettling blind spot.)

If you use an external mouse with macOS, you get a better look at what's going on with scrollbars. I'm on macOS 12.4. Here are some screenshots:

<details><summary>Chrome 101</summary><img width="655" alt="chrome" src="https://user-images.githubusercontent.com/32523293/164651003-f9987dcf-730e-4407-b34d-fa7a7b3d0805.png"></details>

<details><summary>Firefox 99</summary><img width="655" alt="firefox" src="https://user-images.githubusercontent.com/32523293/164652457-19f4ce82-af54-449d-8c3e-65d902c0d7bd.png"></details>

<details><summary>Safari 15.5</summary><img width="655" alt="safari" src="https://user-images.githubusercontent.com/32523293/164652519-3cacce6d-0ca2-4bd4-a58c-906124b48f91.png"></details>

In summary…

1. This behavior can be seen on several platforms, in several browsers. It's not Windows-only or Firefox-only.
2. I still think the problem is caused by the extra (and possibly unnecessary) div that is added when using PDF.js in PDFObject. There's some interaction among the styling of the iframe, the styling of the intermediate div (which is then difficult to change), and the styling of the target node/container.
3. It seems straightforward to fix in PDFObject—you just remove or change a handful of lines in `generatePDFJSMarkup`—but I don't know of a temporary workaround.

**Comment by pipwerks** [comment]
Thanks for all the research everyone, I plan to release an update this weekend.

@theodore-s-beers regarding the extra div, I'm sure there was a reason, but at the moment, I can't recall. It's probably safe to remove but I will look through my notes to sweep out the cobwebs.

**Comment by theodore-s-beers** [comment]
> I'm sure there was a reason, but at the moment, I can't recall.

I was also wondering if it had to do with a use case in which the PDF embed is supposed to take up the whole page. I've never tried that, so I don't know what implications it might have. But there are comments elsewhere in the code referring to the possibility that the target node is the document body. This case is handled in `generatePDFObjectMarkup`, not in `generatePDFJSMarkup`.

**Comment by pipwerks** [comment]
Fixed in 2.2.8, available for testing in a new branch https://github.com/pipwerks/PDFObject/tree/2.2.8

I have tested locally and via BrowserStack, I will merge w Master after I test a little more.

**Comment by dgw** [comment]
Looking forward to this merge. Once I realized that the minified version wasn't updated yet, testing with the unminified script on a couple (desktop) browsers was quite promising. Can't wait to bump production!
